### PR TITLE
Fix Issue #469, Mapper error

### PR DESCRIPTION
**Pull request by nonhermitian**
## Description
Picks a better initial layout for register to physical qubits if none is given.

This is a stopgap fix, as the values should be cached as they need only be computed once.

## Motivation and Context
Multiple issues with swap mapper, as reported in #469.

## How Has This Been Tested?
Seems to solve all mapper issues.

## Types of changes
- [X] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [X] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [X] All new and existing tests passed.

### CODE REVIEW

**Comment by nonhermitian** [comment]
Waiting for @rraymondhp to confirm that his issues have also been resolved by this PR.

File: qiskit/_compiler.py [hunk 1]
```diff
@@ -1,5 +1,7 @@
 # -*- coding: utf-8 -*-
 # pylint: disable=redefined-builtin
+# pylint: disable=C0103
+# pylint: disable=W9006
```
Comment (diego-plan9) [inline qiskit/_compiler.py]: Can you remove this pylint exceptions and take care of the warning themselves instead? In general, disabling a linter check (specially in `qiskit/*`) should only be used as a last resort, not as a convenience, in order to maintain the consistency of the codebase.

File: qiskit/_compiler.py [hunk 2]
```diff
@@ -279,6 +294,54 @@ def load_unroll_qasm_file(filename, basis_gates='u1,u2,u3,cx,id'):
     return circuit_unrolled
 
 
+def best_subset(backend, n_qubits):
```
Comment (diego-plan9) [inline qiskit/_compiler.py]: Would it be possible to have a couple of tests that exercise this function (or that invoke it trough `compile()`)? Probably using the values that were reported in the initial issue or some edge cases would be a good idea.

File: qiskit/_compiler.py [hunk 3]
```diff
@@ -279,6 +294,54 @@ def load_unroll_qasm_file(filename, basis_gates='u1,u2,u3,cx,id'):
...
+    Parameters:
+        backend (Qiskit.BaseBackend): A QISKit backend instance.
+        n_qubits (int): Number of subset qubits to consider.
+
+    Returns:
+        ndarray: Array of qubits to use for best
+                connectivity mapping.
+    """
+    device_qubits = backend.configuration['n_qubits']
+    cmap = np.asarray(backend.configuration['coupling_map'])
```
Comment (atilag) [inline qiskit/_compiler.py]: Can we move this line after all the initial checks? ... so if some of the checks fail, there's no need for this computation.

File: qiskit/_compiler.py [hunk 4]
```diff
@@ -279,6 +294,54 @@ def load_unroll_qasm_file(filename, basis_gates='u1,u2,u3,cx,id'):
...
+
+    if n_qubits > device_qubits:
+        raise ValueError('Number of qubits greater than device.')
+    elif n_qubits <= 0:
+        raise ValueError('Number of qubits <= 0.')
+    if n_qubits == 1:
```
Comment (atilag) [inline qiskit/_compiler.py]: I would put this check as the first line of the method. So if it's true we are returning fast.

File: qiskit/_compiler.py (hunk 2 above)
Comment (nonhermitian) [reply qiskit/_compiler.py]: I am trying to find a short test that raises an error.

File: qiskit/_compiler.py (hunk 1 above)
Comment (nonhermitian) [reply qiskit/_compiler.py]: Fixed the raising exception issue.  Not going to worry about how my iterator constants are defined.

File: qiskit/_compiler.py [hunk 5]
```diff
@@ -279,6 +294,54 @@ def load_unroll_qasm_file(filename, basis_gates='u1,u2,u3,cx,id'):
...
+        return np.array([0])
+
+    data = np.ones_like(cmap[:, 0])
+    sp_cmap = sp.coo_matrix((data, (cmap[:, 0], cmap[:, 1])),
+                            shape=(device_qubits, device_qubits)).tocsr()
+    best = 0
+    best_map = None
+    # do bfs with each node as starting point
+    for kk in range(sp_cmap.shape[0]):
```
Comment (atilag) [inline qiskit/_compiler.py]: Index variables are the only *exception* for one letter naming, doesn't hurt the readability though, but I'd encourage to rename it to just: `k`

File: qiskit/_compiler.py [hunk 6]
```diff
@@ -279,6 +293,57 @@ def load_unroll_qasm_file(filename, basis_gates='u1,u2,u3,cx,id'):
...
+        bfs = cs.breadth_first_order(sp_cmap, i_start=kk, directed=False,
+                                     return_predecessors=False)
+
+        con = 0
```
Comment (atilag) [inline qiskit/_compiler.py]: This is some kind of match counter, right? Can we change the name to something more verbose/explanatory?

File: qiskit/_compiler.py [hunk 7]
```diff
@@ -279,6 +293,57 @@ def load_unroll_qasm_file(filename, basis_gates='u1,u2,u3,cx,id'):
...
+        for ii in range(n_qubits):
+            ss = bfs[ii]
```
Comment (atilag) [inline qiskit/_compiler.py]: Can we have a better naming for `ss`? `bfs_node`? (to distinguish from the `node` of the sparse matrix index)

File: qiskit/_compiler.py [hunk 8]
```diff
@@ -279,6 +293,57 @@ def load_unroll_qasm_file(filename, basis_gates='u1,u2,u3,cx,id'):
...
+
+    Returns:
+        ndarray: Array of qubits to use for best
+                connectivity mapping.
+
+    Raises:
+        ValueError: Wrong number of qubits given.
+    """
+    device_qubits = backend.configuration['n_qubits']
+    cmap = np.asarray(backend.configuration['coupling_map'])
```
Comment (atilag) [inline qiskit/_compiler.py]: Can we move this line before all the checks? so if something is failing in the preconditions there's no need to compute this.

File: qiskit/_compiler.py [hunk 9]
```diff
@@ -279,6 +294,54 @@ def load_unroll_qasm_file(filename, basis_gates='u1,u2,u3,cx,id'):
...
+
+    Returns:
+        ndarray: Array of qubits to use for best
+                connectivity mapping.
+    """
+    device_qubits = backend.configuration['n_qubits']
+    cmap = np.asarray(backend.configuration['coupling_map'])
+
+    if n_qubits > device_qubits:
+        raise ValueError('Number of qubits greater than device.')
```
[the remark this answers is not in the document]
Comment (nonhermitian) [reply qiskit/_compiler.py]: Done.

File: qiskit/_compiler.py [hunk 10]
```diff
@@ -279,6 +294,54 @@ def load_unroll_qasm_file(filename, basis_gates='u1,u2,u3,cx,id'):
...
+    elif n_qubits <= 0:
+        raise ValueError('Number of qubits <= 0.')
```
[the remark this answers is not in the document]
Comment (nonhermitian) [reply qiskit/_compiler.py]: Done.

File: qiskit/_compiler.py (hunk 4 above)
Comment (nonhermitian) [reply qiskit/_compiler.py]: Done.

File: qiskit/_compiler.py (hunk 6 above)
Comment (nonhermitian) [reply qiskit/_compiler.py]: Done.

File: qiskit/_compiler.py (hunk 7 above)
Comment (nonhermitian) [reply qiskit/_compiler.py]: Done.

File: qiskit/_compiler.py (hunk 8 above)
Comment (nonhermitian) [reply qiskit/_compiler.py]: Done.

File: qiskit/_compiler.py (hunk 2 above)
Comment (nonhermitian) [reply qiskit/_compiler.py]: Done.
